SOLR-10654: Introduce output of Prometheus metrics directly from Solr

mlbiscoc opened the pull request:
https://issues.apache.org/jira/browse/SOLR-10654

# Description

Introduce a new Prometheus response writer `/solr/admin/metrics?wt=prometheus` to output Dropwizards Solr Core registry metrics directly from Solr. Output looks similar to the prometheus exporter. This exports `solr.core.*` `solr.node` `solr.jvm`  `solr.jetty` metric registries 

# Solution

Currently, Solr metrics are collected by Dropwizard and differentiated by metric names. This is an anti-pattern to Prometheus. Manually export Dropwizard Solr core registry metrics to a new `SolrPrometheusExporter` base class to parse out the labels from the metric name and output in a Prometheus readable format for the Prometheus Response writer. The exporter is just temporarily created to export a snapshot of what the current Dropwizard registry is at. These are not held in memory.

The exporter will categorize these metrics and export to prometheus format with labels depending on the category and metric names.

In standalone, all labels for the `solr.core` registry includes`core`. In cloud mode, all metrics include `core`, `collection`, `shard`, `replica`

**NOTE: Not all metrics are possible to convert to Prometheus such as Dropwizard string metrics. This is because Prometheus does not support non-numeric types**

# Tests

Added `MetricsHandlerTest`, `SolrPrometheusExporterTest`, `TestPrometheusReponseWriter`

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/solr/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [x] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

richardmlu commented on solr/core/src/java/org/apache/solr/metrics/prometheus/core/SolrCoreSearcherMetric.java:
```diff
@@ -0,0 +1,59 @@
+  public SolrCoreMetric parseLabels() {
+    String[] parsedMetric = metricName.split("\\.");
+    if (dropwizardMetric instanceof Gauge) {
+      String type = parsedMetric[2];
```
Is it possible for `metricName` to be size 1 at this point? If so, we should surround this in a try/throw or a check.

mlbiscoc replied:
I have not seen a metric with delimited with a size of 1 yet but in cases that does, I did surround with try/catch at a higher level [here](https://github.com/apache/solr/blob/d594f50761ec835fedc632ba8d19521918079e05/solr/core/src/java/org/apache/solr/util/stats/MetricUtils.java#L211)

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/core/SolrCoreMetric.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.Map;
+import org.apache.solr.metrics.prometheus.SolrPrometheusCoreRegistry;
+
+public abstract class SolrCoreMetric {
```
I know this is a draft but even a one sentence Javadoc would be helpful.  At a glance, this is Prometheus specific.

mlbiscoc replied:
Will add Javadocs across the new classes

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/core/package-info.java:
```diff
@@ -0,0 +1,22 @@
+ */
+
+/**
+ * The {@link org.apache.solr.metrics.prometheus.core.SolrCoreMetric} is a wrapper to export {@link
```
Perfect; say this on SolrCoreMetric itself

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/SolrPrometheusRegistry.java:
```diff
@@ -0,0 +1,144 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public abstract class SolrPrometheusRegistry {
```
What is the lifecycle of these?  In other words, are they created temporarily to serve a metrics request, or are they lived longer?  I'm a little concerned about synchronization if it lives longer (i.e. when a core goes away, etc.)

mlbiscoc replied:
They are just created temporarily to serve metrics requests so keeping anything in sync will not be an issue. Dropwizard holds all the metrics. We just export the registry to Prometheus when called.

mlbiscoc replied:
Changed name from PrometheusRegistry to Exporter to better understand we are just exporting/transforming dropwizard to prometheus instead of keeping separate registries.

dsmiley reviewed (commented):
There are enough additional dependencies that I feel strongly this should be another module.

It's a shame to see so much code to do the mapping but... shrug, ah well.

mlbiscoc commented:
Moving this out of draft for review

dsmiley commented:
> There are enough additional dependencies that I feel strongly this should be another module.

WDYT?

dsmiley commented:
```
io.prometheus:prometheus-metrics-config:1.1.0 (1 constraints: 3e164ed8)
io.prometheus:prometheus-metrics-exposition-formats:1.1.0 (1 constraints: 0405f335)
io.prometheus:prometheus-metrics-model:1.1.0 (2 constraints: 411b133b)
io.prometheus:prometheus-metrics-shaded-protobuf:1.1.0 (1 constraints: 
```

No, there are 4 dependencies.  Only the last one concerns me.  Why is protobuf needed at all?  Can it be excluded?  Can any of the others be excluded as well, while you're at it?

mlbiscoc commented:
> ```
> io.prometheus:prometheus-metrics-config:1.1.0 (1 constraints: 3e164ed8)
> io.prometheus:prometheus-metrics-exposition-formats:1.1.0 (1 constraints: 0405f335)
> io.prometheus:prometheus-metrics-model:1.1.0 (2 constraints: 411b133b)
> io.prometheus:prometheus-metrics-shaded-protobuf:1.1.0 (1 constraints: 
> ```
> 
> No, there are 4 dependencies. Only the last one concerns me. Why is protobuf needed at all? Can it be excluded? Can any of the others be excluded as well, while you're at it?

Ah ok you were correct. I went ahead and removed the transitive dependencies `prometheus-metrics-shaded-protobuf` and `prometheus-metrics-config` so there should now only be 2

mlbiscoc commented:
@dsmiley or anyone else interested in reviewing. Was curious for more feedback on approach. Is this going in the correct direction or do some disagree if this is correct implementation path towards prometheus in Solr.

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/core/SolrCoreSearcherMetric.java:
```diff
@@ -0,0 +1,70 @@
+  }
+
+  @Override
+  public void toPrometheus(SolrPrometheusCoreExporter solrPrometheusCoreRegistry) {
```
these var names should be changed everywhere so that it's not a "registry"

mlbiscoc replied:
Thanks for catching this. Mistakenly did not change the var names after changing the classes name. Should all be changed now.

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/core/SolrCoreMetric.java:
```diff
@@ -0,0 +1,54 @@
+    this.metricName = metricName;
+    labels.put("core", coreName);
+    if (cloudMode) {
+      String[] coreNameParsed = coreName.split("_");
+      labels.put("collection", coreNameParsed[1]);
```
I could see adding some resiliency to this such that underscores in the collection name still work

mlbiscoc replied:
Good catch. I created a new function with this regex pattern `"^core_(.*)_(shard[0-9]+)_(replica_n[0-9]+)"` for core names looking like `core_test_collection_shard1_replica_n1`. Also created a test under `SolrCoreMetricTest` to avoid regressions if for some reason the pattern for core name changes.

dsmiley commented on solr/core/src/java/org/apache/solr/util/stats/MetricUtils.java:
```diff
@@ -165,6 +169,71 @@ public static void toSolrInputDocuments(
+   * @param compact use compact representation for counters and gauges.
+   * @param consumer consumer that accepts produced {@link SolrPrometheusCoreExporter}-s
+   */
+  public static void toPrometheusRegistry(
```
It's debatable that this belongs here; it doesn't seem utility-like.  It's called by one place, MetricsHandler, so that would be a decent spot.  Alternatively, given its specific-ness to Prometheus, perhaps the PrometheusResponseWriter is a better home.

mlbiscoc replied:
That's a fair point. Moved it to the PrometheusResponseWriter so it's known it's only used for this.

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/SolrPrometheusCoreExporter.java:
```diff
@@ -0,0 +1,104 @@
+  }
+
+  private SolrCoreMetric categorizeCoreMetric(Metric dropwizardMetric, String metricName) {
+    String metricCategory = metricName.split("\\.")[0];
```
I'd rather we don't split merely to get the first component. 
An alternative approach here is to have many startsWith checks so that we're not even parsing/extracting.  I leave it to you to pick a path; either are fine I think.

mlbiscoc replied:
Personally more a fan of split instead of many `startsWith`. I'll keep as if unless someone feels strongly otherwise.

dsmiley replied:
split takes a "limit" second arg.

mlbiscoc replied:
Set limit to 2 and only grab the `[0]` index.

dsmiley commented on solr/core/src/java/org/apache/solr/response/PrometheusResponseWriter.java:
```diff
@@ -0,0 +1,46 @@
+import org.apache.solr.request.SolrQueryRequest;
+
+@SuppressWarnings(value = "unchecked")
+public class PrometheusResponseWriter extends RawResponseWriter {
```
Needs javadocs.  Like say that this is *ONLY* used by the MetricsHandler

mlbiscoc replied:
Done!

dsmiley commented on solr/core/src/java/org/apache/solr/core/SolrCore.java:
```diff
@@ -3032,6 +3034,7 @@ public PluginBag<QueryResponseWriter> getResponseWriters() {
     m.put("csv", new CSVResponseWriter());
     m.put("schema.xml", new SchemaXmlResponseWriter());
     m.put("smile", new SmileResponseWriter());
+    m.put(PROMETHEUS_METRICS_WT, new PrometheusResponseWriter());
```
This is so special-purpose that I think it doesn't belong here.  Can we get away with not registering it?

mlbiscoc replied:
Could not figure out a way to do with without registering it. Is there some example somewhere where a response writer exists without registering the class?

dsmiley replied:
Nevermind.  We even have schema xml format, another special purpose one.  Maybe some day this could be improved.

dsmiley replied:
I filed https://issues.apache.org/jira/browse/SOLR-17354 for improving this matter.

mlbiscoc commented on solr/core/src/java/org/apache/solr/util/stats/MetricUtils.java:
```diff
@@ -165,6 +169,71 @@ public static void toSolrInputDocuments(
+        false,
+        (metricName, metric) -> {
+          try {
+            Metric dropwizardMetric = dropwizardMetrics.get(metricName);
```
A bit confused where you would prefer this. The metric is retrieved within the scope of the `BiConsumer` from `toMap`. I can't move it where `SolrPrometheusCoreExporter` is declared because it is out of scope of this block and the metrics are looped through the registry from `toMaps`.

mlbiscoc commented on solr/core/src/java/org/apache/solr/metrics/prometheus/core/SolrCoreSearcherMetric.java:
```diff
@@ -0,0 +1,70 @@
+        solrPrometheusCoreRegistry.exportGauge(
+            CORE_CACHE_SEARCHER_METRICS,
+            (Gauge<?>) dropwizardMetric,
+            Labels.of(new ArrayList<>(labels.keySet()), new ArrayList<>(labels.values())));
```
Addressed with a `getLabels()` and I agree, it is much cleaner.

mlbiscoc commented on solr/core/src/java/org/apache/solr/metrics/prometheus/SolrPrometheusCoreExporter.java:
```diff
@@ -0,0 +1,104 @@
+  public static final String HIGHLIGHTER = "HIGHLIGHTER";
+  public static final String INDEX = "INDEX";
+  public static final String CORE = "CORE";
+
```
I created `PrometheusExporterInfo` that holds all these constant strings. Let me know what you think.

mlbiscoc commented on solr/core/src/java/org/apache/solr/metrics/prometheus/core/SolrCoreMetric.java:
```diff
@@ -0,0 +1,54 @@
+  public Metric dropwizardMetric;
+  public String coreName;
+  public String metricName;
+  public Map<String, String> labels = new HashMap<>();
```
The `PrometheusTextFormatWriter` will take the MetricSnapshots exported and just output in lexicographical order. So a `HashMap` to `LinkedHashMap` wouldn't really change anything.

mlbiscoc commented on solr/core/src/java/org/apache/solr/response/PrometheusResponseWriter.java:
```diff
@@ -0,0 +1,46 @@
+      throws IOException {
+    NamedList<Object> prometheusRegistries =
+        (NamedList<Object>) response.getValues().get("metrics");
+    Map<String, Object> registryMap = prometheusRegistries.asShallowMap();
```
Got it, moved to an `Iterator` instead.

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/core/SolrCoreSearcherMetric.java:
```diff
@@ -47,24 +45,18 @@ public SolrCoreMetric parseLabels() {
 
   @Override
-  public void toPrometheus(SolrPrometheusCoreExporter solrPrometheusCoreRegistry) {
+  public void toPrometheus(SolrPrometheusCoreExporter solrPrometheusCoreExporter) {
```
nice.  You also could have simply called it "exporter" or "prometheus".  It might then be so short enough  that the usages become one-liners.  Any way it's fine.

mlbiscoc replied:
Changed to just exporter. It still makes sense, I was always just big on verbosity but shorter names for one-liners is better.

dsmiley commented on solr/core/src/java/org/apache/solr/response/PrometheusResponseWriter.java:
```diff
@@ -16,31 +16,115 @@
-    PrometheusTextFormatWriter prometheusTextFormatWriter = new PrometheusTextFormatWriter(false);
-    registryMap.forEach(
-        (name, registry) -> {
+    Iterator<Map.Entry<String, Object>> it = prometheusRegistries.iterator();
```
Using an Iterator manually reminds me of the Java 1.1 days.... in a bad way.  It's needless here, can use Java 7 for loop syntax on any Iterable., saving you 2 lines.

mlbiscoc replied:
Changed to for loop syntax

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/PrometheusExporterInfo.java:
```diff
@@ -0,0 +1,33 @@
+
+public interface PrometheusExporterInfo {
+  /** Category of prefix Solr Core dropwizard handler metric names */
+  enum CoreCategory {
```
Should this go into SolrCoreMetric as it's CORE specific?

mlbiscoc replied:
Implemented it on `SolrPrometheusCoreExporter` instead. It is `CORE` specific so you are correct. Doesn't make sense implementing on the base class.

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/SolrPrometheusCoreExporter.java:
```diff
@@ -0,0 +1,94 @@
+      case QUERY:
+      case UPDATE:
+      case REPLICATION:
+        {
```
can these redundant brackets in these cases be removed?

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/SolrPrometheusExporter.java:
```diff
@@ -0,0 +1,203 @@
+   * Returns an immutable {@link MetricSnapshots} from the {@link
+   * io.prometheus.metrics.model.snapshots.DataPointSnapshot}s collected from the registry
+   */
+  public MetricSnapshots collect() {
```
Does the order matter, like will we return all counters then all gauges in the result?  If so, I could see that being kind of annoying but not a deal breaker of course.

mlbiscoc replied:
Order doesn't matter. The constructor of `MetricSnapshots` gets sorted by the Metric name. See this [constructor](https://github.com/prometheus/client_java/blob/c0a38270c021e4d63998a98a68bacf2b6f32ace5/prometheus-metrics-model/src/main/java/io/prometheus/metrics/model/snapshots/MetricSnapshots.java#L40). But with metrics, it shouldn't matter anyways although I agree, it can be annoying. All the metrics would be aggregated on a Grafana level with prometheus queries.

mlbiscoc commented on solr/core/src/java/org/apache/solr/metrics/prometheus/core/SolrCoreHandlerMetric.java:
```diff
@@ -43,47 +42,35 @@ public SolrCoreMetric parseLabels() {
-    labels.put("category", category);
-    labels.put("type", type);
-    labels.put("handler", handler);
+    labels.putAll(Map.of("category", category, "type", type, "handler", handler));
```
haha duh... 🤦 Changed back accordingly and inline'd them!

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/PrometheusCoreExporterInfo.java:
```diff
@@ -30,4 +32,7 @@ enum CoreCategory {
     CORE
   }
+
+  String PROMETHEUS_CLOUD_CORE_REGEX = "^core_(.*)_(shard[0-9]+)_(replica_n[0-9]+)$";
```
Why separate this into a constant?

dsmiley commented on solr/core/src/java/org/apache/solr/response/PrometheusResponseWriter.java:
```diff
@@ -54,11 +54,11 @@ public void write(OutputStream out, SolrQueryRequest request, SolrQueryResponse
-    while (it.hasNext()) {
-      Map.Entry<String, Object> entry = it.next();
-      SolrPrometheusCoreExporter prometheusExporter = (SolrPrometheusCoreExporter) entry.getValue();
+    for (Iterator<Map.Entry<String, Object>> iter = prometheusRegistries.iterator();
```
This isn't what I meant.  If you use IntelliJ, it should give you a hint -- "'for' loop can be replaced with enhanced 'for' ".  Do that and I think you'll appreciate the results.

mlbiscoc replied:
Sorry was never they best at Java. The enhanced for from the hint is a lot better

dsmiley commented on solr/core/src/java/org/apache/solr/response/PrometheusResponseWriter.java:
```diff
@@ -54,11 +54,11 @@ public void write(OutputStream out, SolrQueryRequest request, SolrQueryResponse
-      SolrPrometheusCoreExporter prometheusExporter = (SolrPrometheusCoreExporter) entry.getValue();
+    for (Iterator<Map.Entry<String, Object>> iter = prometheusRegistries.iterator();
+        iter.hasNext(); ) {
+      SolrPrometheusCoreExporter prometheusExporter =
```
Another use of "var" here is plenty clear enough and will make this a one-liner

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/PrometheusCoreExporterInfo.java:
```diff
@@ -30,4 +32,7 @@ enum CoreCategory {
   }
+
+  String PROMETHEUS_CLOUD_CORE_REGEX = "^core_(.*)_(shard[0-9]+)_(replica_n[0-9]+)$";
+  Pattern PROMETHEUS_CLOUD_CORE_PATTERN = Pattern.compile(PROMETHEUS_CLOUD_CORE_REGEX);
```
But isn't the word "PROMETHEUS" in here actually inaccurate?  This is parsing existing internal names, it's not an external prometheus exported pattern.  I think you could just drop this part.

mlbiscoc replied:
You're right dropped `PROMETHEUS` in the prefix and also removed the `PROMETHEUS_CLOUD_CORE_REGEX` constant

dsmiley reviewed (commented):
Much better.
I know I've been a bit picky/detailed but I'm trying to achieve a level of clarity & conciseness here that will end up being repeated for the other metrics.

What do you think the next step should be?  Totally completing the core metrics so that you have parity with the default Prometheus Exporter config?

mlbiscoc commented:
> Much better. I know I've been a bit picky/detailed but I'm trying to achieve a level of clarity & conciseness here that will end up being repeated for the other metrics.
> 
> What do you think the next step should be? Totally completing the core metrics so that you have parity with the default Prometheus Exporter config?

No worries, I'd prefer you be picky than someone in the future seeing this and it makes no sense to them. I would say next steps are to export the other 3 registries `solr.jetty`, `solr.jvm`, `solr.node` and yes getting as close to the prometheus exporters config as possible. Luckily, in my opinion this was the hardest registry (hopefully I was right) and the ground work for this is already in place to do the rest.

dsmiley commented:
What's here looks good to me.  Just need to add more; lots more metrics.  Maybe it'll evolve a little once you add another category.  My interest on this at-work is not the core metrics but all the other ones (Solr, JVM, Jetty).  I'd then be inclined to take this to our Solr fork to kick the tires for real world usage.

mlbiscoc commented:
> What's here looks good to me. Just need to add more; lots more metrics. Maybe it'll evolve a little once you add another category. My interest on this at-work is not the core metrics but all the other ones (Solr, JVM, Jetty). I'd then be inclined to take this to our Solr fork to kick the tires for real world usage.

Thanks. I've been caught up in some other things recently but plan on continuing on the other registries somewhat soon and it shouldn't be as bad to export. Would it be better for me to continue and commit them to this branch, or was this planned to be merged and export the other 3 registries in a separate PR?

dsmiley commented:
It doesn't feel right to merge this PR right now.  Needs docs.  Also the choice of what metrics are done so far is inverted from the value that would make most sense for this whole approach -- OS/JVM/node issues are highest value.

mlbiscoc commented:
> It doesn't feel right to merge this PR right now. Needs docs. Also the choice of what metrics are done so far is inverted from the value that would make most sense for this whole approach -- OS/JVM/node issues are highest value.

@dsmiley Just finished up exporting `solr.node` `solr.jvm` and `solr.jetty` metrics to prometheus and is ready for another round of review whenever you are. I updated the description above and attached files if you want to see what the prometheus output looks like. Last thing I did not finish writing up is docs. I'm assuming the 2 places to update would be the [metrics-api](https://solr.apache.org/guide/solr/latest/deployment-guide/metrics-reporting.html#metrics-api) section and possibly [Monitoring with Prometheus and Grafana](https://solr.apache.org/guide/solr/latest/deployment-guide/monitoring-with-prometheus-and-grafana.html)?

mlbiscoc commented:
Added ref-guide documentation on Prometheus endpoint

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/jvm/SolrJvmBuffersMetric.java:
```diff
@@ -0,0 +1,52 @@
+
+  @Override
+  public SolrJvmMetric parseLabels() {
+    String[] parsedMetric = metricName.split("\\.");
```
When parsing stuff, it really helps to have a one-line comment of a sample value

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/node/SolrNodeContainerMetric.java:
```diff
@@ -0,0 +1,53 @@
+import org.apache.solr.metrics.prometheus.SolrMetric;
+import org.apache.solr.metrics.prometheus.SolrPrometheusExporter;
+
+/* Dropwizard metrics of name CONTAINER.* */
```
Great; please add similar javadocs for other classes where you forgot

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/node/SolrNodeHandlerMetric.java:
```diff
@@ -0,0 +1,64 @@
+
+/* Dropwizard metrics of name ADMIN.* and UPDATE.* */
+public class SolrNodeHandlerMetric extends SolrNodeMetric {
+  public static final String NODE_REQUESTS = "solr_metrics_node_requests";
+  public static final String NODE_SECONDS_TOTAL = "solr_metrics_node_requests_time";
+  public static final String NODE_CONNECTIONS = "solr_metrics_node_connections";
```
Okay but FWIW *IMO* in situations like this class (parsing/exporting) constants like this only diffuse readability (add a needless indirection).  Take it or leave it; your choice.

mlbiscoc replied:
I personally prefer constants so I'd leave it unless someone feels strongly about it.

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/node/SolrPrometheusNodeExporter.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    SolrMetric solrNodeMetric = categorizeMetric(dropwizardMetric, metricName);
+    solrNodeMetric.parseLabels().toPrometheus(this);
```
if we forget to call parseLabels, will we recognize a mistake?  not a big deal but just a thought

mlbiscoc replied:
I wouldn't say its a mistake. I split these into 2 abstract functions to split out the logic for getting the labels, then converting the correct dropwizard type to the correct prometheus type instead of 1 big function.

I think it's fine if not called. If down the line additional metrics want to be exported from dropwizard, it's possible to just have it exported without any labels and just it's value to which parselabels isn't needed.

dsmiley commented on solr/core/src/java/org/apache/solr/metrics/prometheus/node/SolrPrometheusNodeExporter.java:
```diff
@@ -0,0 +1,99 @@
+
+  private void exportThreadPoolMetric(Metric dropwizardMetric, String metricName) {
+    Labels labels;
+    String[] parsedMetric = metricName.split("\\.");
```
again and in many places, a sample value would be awesome

dsmiley commented on solr/core/src/java/org/apache/solr/response/PrometheusResponseWriter.java:
```diff
@@ -118,13 +108,41 @@ public static void toPrometheus(
+    List<String> parsedRegistry = new ArrayList<>(Arrays.asList(rawParsedRegistry));
+
+    switch (parsedRegistry.get(1)) {
+      case ("core"):
```
These parenthesis in your case statements are pointless; not sure I've ever seen that

dsmiley commented on solr/core/src/java/org/apache/solr/response/PrometheusResponseWriter.java:
```diff
@@ -118,13 +108,41 @@ public static void toPrometheus(
+    String coreName;
+    boolean cloudMode = false;
+    String[] rawParsedRegistry = registryName.split("\\.");
+    List<String> parsedRegistry = new ArrayList<>(Arrays.asList(rawParsedRegistry));
```
why clone?

dsmiley commented on solr/solr-ref-guide/modules/deployment-guide/pages/monitoring-with-prometheus-and-grafana.adoc:
```diff
@@ -587,3 +594,27 @@ This screenshot shows what it might look like:
 .Grafana Dashboard
 image::monitoring-with-prometheus-and-grafana/grafana-solr-dashboard.png[image,width=800]
+
+== Prometheus Metrics API
```
Not sure we want to name it this way... perhaps "Metrics API with Prometheus format"

dsmiley reviewed (commented):
I'd like to suggest an integration test showing all the output (say to a file) stripped of numbers (to remove brittle/changing stuff) to show that we export what we mean to export.  I could imagine someone renaming some metric and then forgetting to update the code here.

mlbiscoc commented:
> I'd like to suggest an integration test showing all the output (say to a file) stripped of numbers (to remove brittle/changing stuff) to show that we export what we mean to export. I could imagine someone renaming some metric and then forgetting to update the code here.

Thanks for reviewing again. Added in an integration test with a file stripped of values. Also the unit tests I wrote originally should catch if any changes that happen to the naming of the prometheus metric or labels.

Also thanks for catching missing javadocs. Made sure to add them in for each class and added sample values for what is being exported to prometheus.

cpoerschke commented on solr/solr-ref-guide/modules/deployment-guide/pages/monitoring-with-prometheus-and-grafana.adoc:
```diff
@@ -16,13 +16,20 @@
 
 A Prometheus exporter (`solr-exporter`) allows users to monitor not only Solr metrics which come from the xref:metrics-reporting.adoc#metrics-api[Metrics API], but also facet counts which come from xref:query-guide:faceting.adoc[] and responses to xref:configuration-guide:collections-api.adoc[] commands and xref:ping.adoc[] requests.
 
+The Metrics API provides a Prometheus Response Writer to output Solr metrics natively to be scraped. It is more efficient and drops the need of running the Prometheus Exporter but at the cost of a fixed output and not as flexiable in terms of configurability.  
```
```suggestion
The Metrics API provides a Prometheus Response Writer to output Solr metrics natively to be scraped. It is more efficient and drops the need of running the Prometheus Exporter but at the cost of a fixed output and not as flexible in terms of configurability.  
```

dsmiley reviewed (commented):
I tried this PR locally using the "techproducts" example.  I thing I noticed is that the metrics filtering appears to not work, which is a big problem.  A test should be added for this, and it should work.  

http://localhost:8983/solr/admin/metrics?category=solr.jetty&wt=prometheus

mlbiscoc commented:
> I tried this PR locally using the "techproducts" example. I thing I noticed is that the metrics filtering appears to not work, which is a big problem. A test should be added for this, and it should work.
> 
> http://localhost:8983/solr/admin/metrics?category=solr.jetty&wt=prometheus

Filtering does work I think the parameter you are looking for is `group`. Try `localhost:8983/solr/admin/metrics?group=jvm&wt=prometheus` and you can change it with `jetty` `node` `core` for the corresponding registries

But still you are correct, I should have a test for this. Will add one in. 

The only other thing I remember now about this is that `regex`, `key`, `expr` parameters for filtering will not work with this prometheus `wt`. I did that on purpose because those 3 parameters are for filtering based on the name of the metric while the others are kind of more generic such as registry or type filtering. I didn't think it made sense because prometheus naming is different from the dropwizard naming after exporting. For example: `solr\.core\..*:QUERY\..*\.requestTimes:max_ms` is not an output of prometheus but to allow `expr` filtering around that felt odd.

Even then if you need extensive filtering or trying to get some specific metric, maybe the Prometheus Exporter process with a custom config might be the better solution.

dsmiley commented:
Oops; maybe I was hurried.  I'm glad the category works at least!  The other filtering options could be deferred to a future JIRA by matching the prometheus name.  It's a very useful capability; some metrics implemented as gauges are expensive to compute, and others are just excessive output that aren't needed for O(N) for N cores if N is for some users quite high.

mlbiscoc commented on solr/core/src/java/org/apache/solr/metrics/prometheus/core/SolrCoreHandlerMetric.java:
```diff
@@ -0,0 +1,77 @@
+    return this;
+  }
+
+  /*
```
Sorry for this. Went with your other suggestion of putting it above `parseLabels` and add an empty line below it.

If you still prefer just normal inline comments, I can make the appropriate changes.

mlbiscoc commented on solr/core/src/java/org/apache/solr/metrics/prometheus/node/SolrNodeHandlerMetric.java:
```diff
@@ -0,0 +1,68 @@
+   */
+  @Override
+  public void toPrometheus(SolrPrometheusExporter exporter) {
+    if (dropwizardMetric instanceof Meter) {
```
I went ahead and removed most of the `instanceof` checks but kept in some places. In the places I kept, it seemed cleaner to export that way as not all my checks are just straight names checks and we don't export every metric. So we can just export a large amount by just checking its type. I can try refactoring further of removing `instanceof` checks if you still disagree.

mlbiscoc commented on solr/core/src/test-files/solr/solr-prometheus-output.txt:
Created it's own prometheus directory to place the file in a place we can place future test files relating to prometheus

mlbiscoc commented on solr/core/src/test/org/apache/solr/response/TestPrometheusResponseWriter.java:
```diff
@@ -0,0 +1,103 @@
+          Files.readString(
+              new File(TEST_PATH().toString(), "solr-prometheus-output.txt").toPath(),
+              StandardCharsets.UTF_8);
+      assertEquals(expectedOutput, actual.replaceAll("(?<=}).*", ""));
```
Parsed out the JVM item tag from the test with additional regex

mlbiscoc commented on solr/core/src/test/org/apache/solr/response/TestPrometheusResponseWriter.java:
```diff
@@ -0,0 +1,103 @@
+          Files.readString(
+              new File(TEST_PATH().toString(), "solr-prometheus-output.txt").toPath(),
+              StandardCharsets.UTF_8);
+      assertEquals(expectedOutput, actual.replaceAll("(?<=}).*", ""));
```
Added in a comment with explanation.

mlbiscoc commented:
> Oops; maybe I was hurried. I'm glad the category works at least! The other filtering options could be deferred to a future JIRA by matching the prometheus name. It's a very useful capability; some metrics implemented as gauges are expensive to compute, and others are just excessive output that aren't needed for O(N) for N cores if N is for some users quite high.

Added in a test for the filtering. I'd be happy expand on this and add further filtering feature specific to prometheus after this is ready to be merged.

dsmiley commented on solr/core/src/test/org/apache/solr/handler/admin/MetricsHandlerTest.java:
```diff
@@ -677,6 +706,340 @@ public void testExprMetrics() throws Exception {
+
+    actualSnapshot = getMetricSnapshot(actualSnapshots, "solr_metrics_jvm_gc_seconds");
+    actualGaugeDataPoint =
+        getGaugeDatapointSnapshot(actualSnapshot, Labels.of("item", "G1-Old-Generation"));
```
This assertion will fail for varying JVMs; skip this one.  Some others too above... it's okay to make the test looser / less specific.

mlbiscoc replied:
Removed a few more tests from the JVM registry

dsmiley commented on solr/solr-ref-guide/modules/deployment-guide/pages/metrics-reporting.adoc:
```diff
@@ -879,7 +879,7 @@ The response will look like this:
 ----
 
-Like other request handlers, the Metrics API can also take the `wt` parameter to define the output format.
+Like other request handlers, the Metrics API can also take the `wt` parameter to define the output format. The Metrics API also provides an additional special `wt` parameter `prometheus` that will output metrics in Prometheus format. This can be used for xref:monitoring-with-prometheus-and-grafana.adoc#metrics-api-with-prometheus-format[Monitoring with Prometheus and Grafana]
```
After each sentence, put a newline (doesn't change the final render but it to keep editing & diffs more manageable).

Should document what options *do not work* with prometheus.  The regex one doesn't, you said.

mlbiscoc replied:
Added in a small note about the filter query parameters. Let me know if you prefer a different way to write it on the ref-guide

dsmiley reviewed (commented):
This is super close now.
Can you propose a brief CHANGES.txt under 9.7 "New Features" (I think it's closer to this than an improvement).

mlbiscoc commented:
> This is super close now. Can you propose a brief CHANGES.txt under 9.7 "New Features" (I think it's closer to this than an improvement).

Added entry into CHANGES.txt

dsmiley commented on solr/core/src/test/org/apache/solr/response/TestPrometheusResponseWriter.java:
```diff
@@ -0,0 +1,105 @@
+    solrClientTestRule.startSolr(LuceneTestCase.createTempDir());
+    jetty = solrClientTestRule.getJetty();
+    solrClientTestRule.newCollection().withConfigSet(ExternalPaths.DEFAULT_CONFIGSET).create();
+    jetty.getCoreContainer().waitForLoadingCoresToFinish(30000);
```
@mlbiscoc Why did you add this line to wait 30secs?  The [default](https://github.com/apache/solr/blob/2c55000296bb04e12e2bfae83eca7e5f11caba0e/solr/test-framework/src/java/org/apache/solr/embedded/JettyConfig.java#L85) is 5min which is plenty.  Creating a core should be synchronous.

@epugh probably copied you in a recent PR and I had the same question.

epugh replied:
I have fixed this in #3988 after auditing places were .waitForLoadingCoresToFinish is called...

mlbiscoc replied:
Hmm I don't remember as this was almost 2 years ago... But I know I would steal templates from other tests as well and it looks like `MetricsHandlerTest` does the same which might be why I did it.
